File: bridge/runtimes.py
from __future__ import annotations

import asyncio
import os
import re
import tempfile
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from copilot import CopilotClient
from copilot._cli_download import get_or_download_cli
from copilot.generated.rpc import AccountGetQuotaRequest

from openai_payload import completion_payload, prompt_from_payload, stream_chunks
from store import AuthRecord, CredentialStore


URL_PATTERN = re.compile(r"https://[^\s<>\"']+")
USER_CODE_PATTERN = re.compile(r"\b[A-Z0-9]{4}-[A-Z0-9]{4}\b")


@dataclass(slots=True)
class LoginFlow:
    provider: str
    handle: str
    process: asyncio.subprocess.Process
    output: list[str] = field(default_factory=list)
    url: str = ""
    user_code: str = ""
    reader: asyncio.Task[None] | None = None


class SubscriptionRuntimes:
    def __init__(self, store: CredentialStore, cursor_binary: Path) -> None:
        self.store = store
        self.cursor_binary = cursor_binary
        self.logins: dict[str, LoginFlow] = {}

    async def start_login(self, provider: str) -> tuple[str, str, dict[str, Any]]:
        handle = self.store.new_handle()
        state = self.store.new_handle()
        home = self.store.home_dir(provider, handle)
        command, env = self._login_command(provider, home)
        process = await asyncio.create_subprocess_exec(
            *command,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.STDOUT,
            env=env,
        )
        flow = LoginFlow(provider=provider, handle=handle, process=process)
        flow.reader = asyncio.create_task(self._read_login_output(flow))
        self.logins[state] = flow

        for _ in range(300):
            if flow.url:
                url = flow.url
                if provider == "copilot" and flow.user_code:
                    url = f"{url}?user_code={flow.user_code}"
                return url, state, {
                    "instructions": self._recent_output(flow),
                    "user_code": flow.user_code,
                }
            if process.returncode is not None:
                break
            await asyncio.sleep(0.1)
        await self._stop_flow(flow)
        self.logins.pop(state, None)
        raise RuntimeError(f"{provider} login did not produce an authorization URL: {self._recent_output(flow)}")

    async def poll_login(self, provider: str, state: str) -> tuple[str, str, AuthRecord | None]:
        flow = self.logins.get(state)
        if flow is None or flow.provider != provider:
            return "error", "login state was not found or expired", None
        return_code = flow.process.returncode
        if return_code is None:
            return "pending", self._recent_output(flow) or "waiting for browser authorization", None
        if flow.reader is not None:
            await asyncio.gather(flow.reader, return_exceptions=True)
        self.logins.pop(state, None)
        if return_code != 0:
            return "error", self._recent_output(flow) or f"login exited with code {return_code}", None
        record = self.store.save_record(
            provider,
            flow.handle,
            label=f"{provider.title()} subscription {flow.handle[:6]}",
        )
        return "success", "login completed", record

    async def list_models(self, provider: str, handle: str) -> list[dict[str, Any]]:
        self.store.assert_account(provider, handle)
        if provider == "copilot":
            return await self._copilot_models(handle)
        return await self._cursor_models(handle)

    async def execute(
        self,
        provider: str,
        handle: str,
        model: str,
        payload: dict[str, Any],
        stream: bool,
    ) -> dict[str, Any]:
        self.store.assert_account(provider, handle)
        prompt = prompt_from_payload(payload)
        if provider == "copilot":
            content = await self._copilot_execute(handle, model, prompt)
        else:
            content = await self._cursor_execute(handle, model, prompt)
        if stream:
            return {"payload": {}, "chunks": stream_chunks(model, content)}
        return {"payload": completion_payload(model, content)}

    async def quota(self, provider: str, handle: str) -> dict[str, Any]:
        self.store.assert_account(provider, handle)
        if provider == "copilot":
            home = self.store.home_dir(provider, handle)
            client = self._copilot_client(home)
            await client.start()
            try:
                result = await client.rpc.account.get_quota(AccountGetQuotaRequest())
                return result.to_dict()
            finally:
                await client.stop()
        output = await self._run_cursor(handle, ["status"], timeout=30)
        return {"status": output.strip(), "source": "cursor-agent status"}

    def _login_command(self, provider: str, home: Path) -> tuple[list[str], dict[str, str]]:
        env = os.environ.copy()
        env["HOME"] = str(home)
        env["NO_OPEN_BROWSER"] = "1"
        env["BROWSER"] = "echo"
        env["GH_BROWSER"] = "echo"
        if provider == "copilot":
            env["COPILOT_HOME"] = str(home / ".copilot")
            cli = get_or_download_cli()
            if not cli:
                raise RuntimeError("official Copilot CLI runtime is unavailable")
            return [cli, "login"], env
        if provider == "cursor":
            return [str(self.cursor_binary), "login"], env
        raise ValueError("unsupported provider")

    async def _read_login_output(self, flow: LoginFlow) -> None:
        if flow.process.stdout is None:
            return
        while True:
            line = await flow.process.stdout.readline()
            if not line:
                break
            text = line.decode("utf-8", errors="replace").strip()
            if text:
                flow.output.append(text)
                match = URL_PATTERN.search(text)
                if match and not flow.url:
                    flow.url = match.group(0).rstrip(".,;)")
                code_match = USER_CODE_PATTERN.search(text)
                if code_match and not flow.user_code:
                    flow.user_code = code_match.group(0)
        await flow.process.wait()

    async def _stop_flow(self, flow: LoginFlow) -> None:
        if flow.process.returncode is None:
            flow.process.terminate()
            try:
                await asyncio.wait_for(flow.process.wait(), timeout=3)
            except TimeoutError:
                flow.process.kill()
                await flow.process.wait()
        if flow.reader is not None:
            await asyncio.gather(flow.reader, return_exceptions=True)

    @staticmethod
    def _recent_output(flow: LoginFlow) -> str:
        return "\n".join(flow.output[-12:])[-4000:]

    def _copilot_client(self, home: Path) -> CopilotClient:
        env = os.environ.copy()
        env["HOME"] = str(home)
        env["COPILOT_HOME"] = str(home / ".copilot")
        return CopilotClient(
            base_directory=str(home / ".copilot"),
            env=env,
            use_logged_in_user=True,
            mode="empty",
        )

    async def _copilot_models(self, handle: str) -> list[dict[str, Any]]:
        home = self.store.home_dir("copilot", handle)
        client = self._copilot_client(home)
        await client.start()
        try:
            models = await client.list_models()
            return [
                {
                    "id": model.id,
                    "display_name": getattr(model, "name", None) or model.id,
                    "description": getattr(model, "description", None) or "GitHub Copilot model",
                    "context_length": getattr(model, "max_context_window_tokens", None) or 0,
                    "max_completion_tokens": getattr(model, "max_output_tokens", None) or 0,
                }
                for model in models
            ]
        finally:
            await client.stop()

    async def _copilot_execute(self, handle: str, model: str, prompt: str) -> str:
        home = self.store.home_dir("copilot", handle)
        client = self._copilot_client(home)
        await client.start()
        session = None
        try:
            session = await client.create_session(
                model=model or None,
                tools=[],
                available_tools=[],
                working_directory=str(home),
                skip_custom_instructions=True,
            )
            message = await session.send_and_wait(prompt, timeout=170)
            return message.data.content if message is not None else ""
        finally:
            if session is not None:
                await session.disconnect()
            await client.stop()

    async def _cursor_models(self, handle: str) -> list[dict[str, Any]]:
        output = await self._run_cursor(handle, ["--list-models"], timeout=30)
        models: list[dict[str, Any]] = []
        for line in output.splitlines():
            value = line.strip().lstrip("-* ").strip()
            if not value or "model" in value.lower() and ":" not in value:
                continue
            model_id = value.split()[0]
            if re.fullmatch(r"[A-Za-z0-9._:/-]+", model_id):
                models.append({"id": model_id, "display_name": value})
        if not models:
            models.append({"id": "auto", "display_name": "Auto"})
        return models

    async def _cursor_execute(self, handle: str, model: str, prompt: str) -> str:
        with tempfile.TemporaryDirectory(prefix="cpa-cursor-") as workspace:
            args = [
                "--print",
                "--trust",
                "--mode",
                "ask",
                "--workspace",
                workspace,
                "--output-format",
                "text",
            ]
            if model:
                args.extend(["--model", model])
            args.append(prompt)
            return (await self._run_cursor(handle, args, timeout=170)).strip()

    async def _run_cursor(self, handle: str, args: list[str], *, timeout: int) -> str:
        home = self.store.home_dir("cursor", handle)
        env = os.environ.copy()
        env["HOME"] = str(home)
        process = await asyncio.create_subprocess_exec(
            str(self.cursor_binary),
            *args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.STDOUT,
            env=env,
        )
        try:
            stdout, _ = await asyncio.wait_for(process.communicate(), timeout=timeout)
        except TimeoutError:
            process.kill()
            await process.wait()
            raise RuntimeError("Cursor CLI timed out")
        output = stdout.decode("utf-8", errors="replace")
        if process.returncode != 0:
            raise RuntimeError(f"Cursor CLI exited with {process.returncode}: {output[-4000:]}")
        return output

File: internal/bridge/runtimes.go
package bridge

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/rpc"
)

type Runtimes struct {
	store         *CredentialStore
	copilotBinary string
	cursorBinary  string
	cursorScript  string

	mu     sync.Mutex
	logins map[string]*loginFlow
}

func NewRuntimes(store *CredentialStore, copilotBinary, cursorBinary, cursorScript string) *Runtimes {
	return &Runtimes{
		store: store, copilotBinary: copilotBinary, cursorBinary: cursorBinary, cursorScript: cursorScript,
		logins: make(map[string]*loginFlow),
	}
}

func (r *Runtimes) StartLogin(ctx context.Context, provider string) (string, string, map[string]any, error) {
	handle, err := r.store.NewHandle()
	if err != nil {
		return "", "", nil, err
	}
	state, err := r.store.NewHandle()
	if err != nil {
		return "", "", nil, err
	}
	home, err := r.store.HomeDir(provider, handle)
	if err != nil {
		return "", "", nil, err
	}
	command, err := r.loginCommand(provider, home)
	if err != nil {
		return "", "", nil, err
	}
	flow, err := startLoginFlow(provider, handle, command)
	if err != nil {
		return "", "", nil, fmt.Errorf("start %s login: %w", provider, err)
	}
	r.mu.Lock()
	r.logins[state] = flow
	r.mu.Unlock()
	go r.expireLogin(state, flow)

	timeout := time.NewTimer(30 * time.Second)
	defer timeout.Stop()
	ticker := time.NewTicker(100 * time.Millisecond)
	defer ticker.Stop()
	for {
		url, userCode, recent, finished, waitErr := flow.snapshot()
		if url != "" {
			return authorizationURL(provider, url, userCode), state, loginMetadata(url, userCode, recent), nil
		}
		if finished {
			r.removeLogin(state)
			if waitErr == nil {
				waitErr = errors.New("login command exited")
			}
			return "", "", nil, fmt.Errorf("%s login did not produce an authorization URL: %s: %w", provider, recent, waitErr)
		}
		select {
		case <-ctx.Done():
			flow.stop()
			r.removeLogin(state)
			return "", "", nil, ctx.Err()
		case <-timeout.C:
			flow.stop()
			r.removeLogin(state)
			return "", "", nil, fmt.Errorf("%s login did not produce an authorization URL: %s", provider, recent)
		case <-ticker.C:
		}
	}
}

func (r *Runtimes) PollLogin(provider, state string) (string, string, *AuthRecord) {
	r.mu.Lock()
	flow := r.logins[state]
	r.mu.Unlock()
	if flow == nil || flow.provider != provider {
		return "error", "login state was not found or expired", nil
	}
	_, _, recent, finished, waitErr := flow.snapshot()
	if !finished {
		if recent == "" {
			recent = "waiting for browser authorization"
		}
		return "pending", recent, nil
	}
	r.removeLogin(state)
	if waitErr != nil {
		if recent == "" {
			recent = waitErr.Error()
		}
		return "error", recent, nil
	}
	record, err := r.store.SaveRecord(
		provider,
		flow.handle,
		strings.ToUpper(provider[:1])+provider[1:]+" subscription "+flow.handle[:6],
		"",
	)
	if err != nil {
		return "error", err.Error(), nil
	}
	return "success", "login completed", &record
}

func (r *Runtimes) ListModels(ctx context.Context, provider, handle string) ([]map[string]any, error) {
	if err := r.store.AssertAccount(provider, handle); err != nil {
		return nil, err
	}
	if provider == "cursor" {
		return r.cursorModels(ctx, handle)
	}
	client, err := r.copilotClient(provider, handle)
	if err != nil {
		return nil, err
	}
	if err := client.Start(ctx); err != nil {
		return nil, err
	}
	defer client.Stop()
	models, err := client.ListModels(ctx)
	if err != nil {
		return nil, err
	}
	result := make([]map[string]any, 0, len(models))
	for _, model := range models {
		contextLength := 0
		if model.Capabilities.Limits.MaxContextWindowTokens != nil {
			contextLength = *model.Capabilities.Limits.MaxContextWindowTokens
		}
		name := model.Name
		if name == "" {
			name = model.ID
		}
		result = append(result, map[string]any{
			"id": model.ID, "display_name": name, "description": "GitHub Copilot model",
			"context_length": contextLength, "max_completion_tokens": 0,
		})
	}
	return result, nil
}

func (r *Runtimes) Execute(ctx context.Context, provider, handle, model string, payload map[string]any, stream bool) (map[string]any, error) {
	if err := r.store.AssertAccount(provider, handle); err != nil {
		return nil, err
	}
	prompt, err := promptFromPayload(payload)
	if err != nil {
		return nil, err
	}
	var content string
	if provider == "copilot" {
		content, err = r.copilotExecute(ctx, handle, model, prompt)
	} else {
		content, err = r.cursorExecute(ctx, handle, model, prompt)
	}
	if err != nil {
		return nil, err
	}
	if stream {
		chunks, err := streamChunks(model, content)
		if err != nil {
			return nil, err
		}
		return map[string]any{"payload": map[string]any{}, "chunks": chunks}, nil
	}
	return map[string]any{"payload": completionPayload(model, content)}, nil
}

func (r *Runtimes) Quota(ctx context.Context, provider, handle string) (any, error) {
	if err := r.store.AssertAccount(provider, handle); err != nil {
		return nil, err
	}
	if provider == "cursor" {
		output, err := r.runCursor(ctx, handle, []string{"status"}, 30*time.Second)
		if err != nil {
			return nil, err
		}
		return map[string]any{"status": strings.TrimSpace(output), "source": "cursor-agent status"}, nil
	}
	client, err := r.copilotClient(provider, handle)
	if err != nil {
		return nil, err
	}
	if err := client.Start(ctx); err != nil {
		return nil, err
	}
	defer client.Stop()
	return client.RPC.Account.GetQuota(ctx, &rpc.AccountGetQuotaRequest{})
}

func (r *Runtimes) loginCommand(provider, home string) (*exec.Cmd, error) {
	var command *exec.Cmd
	switch provider {
	case "copilot":
		command = exec.Command(r.copilotBinary, "login")
	case "cursor":
		command = exec.Command(r.cursorBinary, r.cursorArguments("login")...)
	default:
		return nil, errors.New("unsupported provider")
	}
	command.Env = accountEnvironment(provider, home)
	return command, nil
}

func (r *Runtimes) copilotClient(provider, handle string) (*copilot.Client, error) {
	home, err := r.store.HomeDir(provider, handle)
	if err != nil {
		return nil, err
	}
	useLoggedInUser := true
	baseDirectory := filepath.Join(home, ".copilot")
	if err := os.MkdirAll(baseDirectory, 0o700); err != nil {
		return nil, err
	}
	return copilot.NewClient(&copilot.ClientOptions{
		Connection:       copilot.StdioConnection{Path: r.copilotBinary, Env: accountEnvironment(provider, home)},
		WorkingDirectory: home,
		BaseDirectory:    baseDirectory,
		UseLoggedInUser:  &useLoggedInUser,
		Mode:             copilot.ModeEmpty,
	}), nil
}

func (r *Runtimes) copilotExecute(ctx context.Context, handle, model, prompt string) (string, error) {
	client, err := r.copilotClient("copilot", handle)
	if err != nil {
		return "", err
	}
	home, err := r.store.HomeDir("copilot", handle)
	if err != nil {
		return "", err
	}
	if err := client.Start(ctx); err != nil {
		return "", err
	}
	defer client.Stop()
	skipInstructions := true
	config := &copilot.SessionConfig{
		ClientName:             "cpa-copilot-cursor",
		Model:                  model,
		Tools:                  []copilot.Tool{},
		AvailableTools:         []string{},
		WorkingDirectory:       home,
		SkipCustomInstructions: &skipInstructions,
	}
	session, err := client.CreateSession(ctx, config)
	if err != nil {
		return "", err
	}
	defer session.Disconnect()
	callContext, cancel := context.WithTimeout(ctx, 170*time.Second)
	defer cancel()
	event, err := session.SendAndWait(callContext, copilot.MessageOptions{Prompt: prompt})
	if err != nil {
		return "", err
	}
	if event == nil {
		return "", nil
	}
	message, ok := event.Data.(*copilot.AssistantMessageData)
	if !ok {
		return "", fmt.Errorf("unexpected Copilot response type %T", event.Data)
	}
	return message.Content, nil
}

func (r *Runtimes) cursorModels(ctx context.Context, handle string) ([]map[string]any, error) {
	output, err := r.runCursor(ctx, handle, []string{"--list-models"}, 30*time.Second)
	if err != nil {
		return nil, err
	}
	pattern := regexp.MustCompile(`^[A-Za-z0-9._:/-]+$`)
	models := make([]map[string]any, 0)
	for _, line := range strings.Split(output, "\n") {
		value := strings.TrimSpace(strings.TrimLeft(strings.TrimSpace(line), "-* "))
		if value == "" || strings.Contains(strings.ToLower(value), "model") && !strings.Contains(value, ":") {
			continue
		}
		modelID := strings.Fields(value)[0]
		if pattern.MatchString(modelID) {
			models = append(models, map[string]any{"id": modelID, "display_name": value})
		}
	}
	if len(models) == 0 {
		models = append(models, map[string]any{"id": "auto", "display_name": "Auto"})
	}
	return models, nil
}

func (r *Runtimes) cursorExecute(ctx context.Context, handle, model, prompt string) (string, error) {
	workspace, err := os.MkdirTemp("", "cpa-cursor-")
	if err != nil {
		return "", err
	}
	defer os.RemoveAll(workspace)
	args := []string{"--print", "--trust", "--mode", "ask", "--workspace", workspace, "--output-format", "text"}
	if model != "" {
		args = append(args, "--model", model)
	}
	args = append(args, prompt)
	output, err := r.runCursor(ctx, handle, args, 170*time.Second)
	return strings.TrimSpace(output), err
}

func (r *Runtimes) runCursor(ctx context.Context, handle string, args []string, timeout time.Duration) (string, error) {
	home, err := r.store.HomeDir("cursor", handle)
	if err != nil {
		return "", err
	}
	callContext, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	command := exec.CommandContext(callContext, r.cursorBinary, r.cursorArguments(args...)...)
	command.Env = accountEnvironment("cursor", home)
	var output bytes.Buffer
	command.Stdout = &output
	command.Stderr = &output
	if err := command.Run(); err != nil {
		return "", fmt.Errorf("Cursor CLI exited: %s: %w", tail(output.String(), 4000), err)
	}
	return output.String(), nil
}

func (r *Runtimes) cursorArguments(args ...string) []string {
	if r.cursorScript == "" {
		return args
	}
	return append([]string{"--use-system-ca", r.cursorScript}, args...)
}

func (r *Runtimes) removeLogin(state string) {
	r.mu.Lock()
	delete(r.logins, state)
	r.mu.Unlock()
}

func (r *Runtimes) expireLogin(state string, flow *loginFlow) {
	timer := time.NewTimer(15 * time.Minute)
	defer timer.Stop()
	<-timer.C

	r.mu.Lock()
	if r.logins[state] != flow {
		r.mu.Unlock()
		return
	}
	delete(r.logins, state)
	r.mu.Unlock()
	flow.stop()
}

func accountEnvironment(provider, home string) []string {
	environment := append([]string{}, os.Environ()...)
	environment = setEnvironment(environment, "HOME", home)
	environment = setEnvironment(environment, "NO_OPEN_BROWSER", "1")
	environment = setEnvironment(environment, "BROWSER", "echo")
	environment = setEnvironment(environment, "GH_BROWSER", "echo")
	if provider == "copilot" {
		environment = setEnvironment(environment, "COPILOT_HOME", filepath.Join(home, ".copilot"))
	} else {
		environment = setEnvironment(environment, "CURSOR_INVOKED_AS", "cursor-agent")
		environment = setEnvironment(environment, "NODE_COMPILE_CACHE", filepath.Join(home, ".cache", "cursor-compile-cache"))
	}
	return environment
}

func setEnvironment(environment []string, name, value string) []string {
	prefix := name + "="
	result := make([]string, 0, len(environment)+1)
	for _, entry := range environment {
		if !strings.HasPrefix(entry, prefix) {
			result = append(result, entry)
		}
	}
	return append(result, prefix+value)
}

func tail(value string, limit int) string {
	if len(value) <= limit {
		return value
	}
	return value[len(value)-limit:]
}
